### ISSUE TITLE
SC2086 requiring quotes for `$?` et al. again

### ISSUE THREAD

**Issue by AloisMahdal** [koalaman/shellcheck#2560]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2086
- My shellcheck version (`shellcheck --version` or "online"): online, and v0.8.0-62-g4806719 (koalaman/shellcheck-alpine in [my GitLab job](https://gitlab.com/vornet/shellfu/shellfu/-/jobs/2884485541))
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/bash

some_fun() {
    (
        foo; es=$?
        echo $es
    )
}

(
    foo; es=$?
    echo $es
)

# some_fun  ## strange: uncomment this and the finding is gone
```

#### Here's what shellcheck currently says:

```
$ shellcheck myscript
 
Line 7:
        echo $es
             ^-- SC2086: Double quote to prevent globbing and word splitting.

```

#### Here's what I wanted or expected to see:

it should be ignored, as `$?` cannot have special chars in it, and `$es` is merely a snapshot of the value

Note that:
 *  SC2086 is not raised for the code that's outside *some_fun()*,
 *  SC2086 is also not raised if I add `some_fun` (call the function) at the end of the script
 *  this is tolerated in v0.7.1
 *  similar case also probably applies: when the variable is set as a constant literal word, eg. `foo=bar; echo $foo`

**Comment by a1346054** [comment]
Not a bug.

The correct behavior is to quote all substitutions, including `$?` because it can indeed interact with things such as IFS=0

**Comment by koalaman** [comment]
It's definitely unexpected that the warning is only emitted when the function is not invoked. In this case it was in fact a bug related the ShellCheck's data flow analysis.

ShellCheck examines the flow of the program, and processes any functions as they're invoked with the context they're invoked in. Any uninvoked functions are in a sense dead, but will in practice be invoked later by a sourcing script. Therefore, ShellCheck processes any uninvoked functions with the state the script would be in after exiting.

The bug was that ShellCheck used the wrong state during this processing.

It should have used the initial state plus any modifications made by the script, but instead it only used the modifications. This means that it effectively forgot that shell variables like `$BASHPID` and `$COLUMNS` are set whether or not the script assigns them, and `$?` is one of those. With `$?` gone, it was assumed to be unknown instead of numeric, causing the issue in uninvoked functions.

This is now fixed, thanks!

(For reminders to quote *all* variables to account for things like IFS=0, enable `quote-safe-variables`)

**Comment by a1346054** [comment]
Using `quote-safe-variables` is not enough, it doesn't produce a reminder to quote *all* variables where they need to be quoted to account for things like IFS=0.

```bash
#!/bin/bash
#shellcheck enable=quote-safe-variables
IFS=0
echo $$
true
echo $?
a=0
echo $a
```

tested on https://shellcheck.net/ only produces
```
$ shellcheck myscript
 
Line 8:
echo $a
     ^-- SC2248 (style): Prefer double quoting even when variables don't contain special characters.

Did you mean: (apply this, apply all SC2248)
echo "$a"
```

and doesn't warn about the unquoted `$$` or `$?`
